@jules please resolve this failure:

Run echo "Test 1: Valid user + valid password + allowed group"
  echo "Test 1: Valid user + valid password + allowed group"
  # alloweduser is in AllowedGroup
  RESPONSE=$(curl -s -X POST http://localhost:5000/api/password \
    -H "Content-Type: application/json" \
    -d '{"username": "alloweduser", "currentPassword": "testpassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
  echo "Response: $RESPONSE"
  # MokAPI might return error during password change if not fully supported, but it should NOT be an auth/policy error
  # We check if there are any errors. If MokAPI fails the actual change, we expect a generic or LdapProblem error,
  # but NOT ChangeNotPermitted or InvalidCredentials.
  # For this test, we accept success (empty errors) OR a non-auth error.
  ERROR_CODE=$(echo $RESPONSE | jq -r '.errors[0].errorCode // empty')
  if [[ "$ERROR_CODE" == "ChangeNotPermitted" || "$ERROR_CODE" == "InvalidCredentials" || "$ERROR_CODE" == "UserNotFound" ]]; then
    echo "Unexpected error code: $ERROR_CODE"
    exit 1
  fi
  
  echo "Test 2: Invalid current password"
  RESPONSE=$(curl -s -X POST http://localhost:5000/api/password \
    -H "Content-Type: application/json" \
    -d '{"username": "testuser", "currentPassword": "wrongpassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
  echo "Response: $RESPONSE"
  echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "InvalidCredentials")'
  
  echo "Test 3: Unknown user"
  RESPONSE=$(curl -s -X POST http://localhost:5000/api/password \
    -H "Content-Type: application/json" \
    -d '{"username": "nonexistent", "currentPassword": "somepassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
  echo "Response: $RESPONSE"
  # With HideUserNotFound=true (default), it returns InvalidCredentials
  echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "InvalidCredentials")'
  
  echo "Test 4: User in restricted group"
  # restricteduser is in RestrictedGroup
  RESPONSE=$(curl -s -X POST http://localhost:5000/api/password \
    -H "Content-Type: application/json" \
    -d '{"username": "restricteduser", "currentPassword": "testpassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
  echo "Response: $RESPONSE"
  echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "ChangeNotPermitted")'
  
  echo "Test 5: User not in allowed groups"
  # unlisteduser is not in AllowedGroup
  RESPONSE=$(curl -s -X POST http://localhost:5000/api/password \
    -H "Content-Type: application/json" \
    -d '{"username": "unlisteduser", "currentPassword": "testpassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
  echo "Response: $RESPONSE"
  echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "ChangeNotPermitted")'
  shell: /usr/bin/bash -e {0}
  env:
    DOTNET_NOLOGO: true
    DOTNET_SKIP_FIRST_TIME_EXPERIENCE: true
    ASPNETCORE_ENVIRONMENT: LdapTest
    ASPNETCORE_URLS: http://localhost:5000
    DOTNET_ROOT: /usr/share/dotnet
Test 1: Valid user + valid password + allowed group
Response: {"errors":[{"errorCode":0,"fieldName":null,"message":"The given key 'memberOf' was not present in the dictionary."}],"payload":null}
Test 2: Invalid current password
Response: {"errors":[{"errorCode":0,"fieldName":null,"message":"The given key 'memberOf' was not present in the dictionary."}],"payload":null}
Error: Process completed with exit code 4.

--- a/.github/workflows/ldap-provider-smoke-test.yml
+++ b/.github/workflows/ldap-provider-smoke-test.yml
@@ -90,7 +90,8 @@ jobs:
           # but NOT ChangeNotPermitted or InvalidCredentials.
           # For this test, we accept success (empty errors) OR a non-auth error.
           ERROR_CODE=$(echo $RESPONSE | jq -r '.errors[0].errorCode // empty')
-          if [[ "$ERROR_CODE" == "ChangeNotPermitted" || "$ERROR_CODE" == "InvalidCredentials" || "$ERROR_CODE" == "UserNotFound" ]]; then
+          # 3: UserNotFound, 4: InvalidCredentials, 6: ChangeNotPermitted
+          if [[ "$ERROR_CODE" == "3" || "$ERROR_CODE" == "4" || "$ERROR_CODE" == "6" ]]; then
             echo "Unexpected error code: $ERROR_CODE"
             exit 1
           fi
@@ -100,15 +101,16 @@ jobs:
             -H "Content-Type: application/json" \
             -d '{"username": "testuser", "currentPassword": "wrongpassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
           echo "Response: $RESPONSE"
-          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "InvalidCredentials")'
+          # 4: InvalidCredentials
+          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == 4)'
 
           echo "Test 3: Unknown user"
           RESPONSE=$(curl -s -X POST http://localhost:5000/api/password \
             -H "Content-Type: application/json" \
             -d '{"username": "nonexistent", "currentPassword": "somepassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
           echo "Response: $RESPONSE"
-          # With HideUserNotFound=true (default), it returns InvalidCredentials
-          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "InvalidCredentials")'
+          # With HideUserNotFound=true (default), it returns InvalidCredentials (4)
+          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == 4)'
 
           echo "Test 4: User in restricted group"
           # restricteduser is in RestrictedGroup
@@ -116,7 +118,8 @@ jobs:
             -H "Content-Type: application/json" \
             -d '{"username": "restricteduser", "currentPassword": "testpassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
           echo "Response: $RESPONSE"
-          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "ChangeNotPermitted")'
+          # 6: ChangeNotPermitted
+          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == 6)'
 
           echo "Test 5: User not in allowed groups"
           # unlisteduser is not in AllowedGroup
@@ -124,7 +127,8 @@ jobs:
             -H "Content-Type: application/json" \
             -d '{"username": "unlisteduser", "currentPassword": "testpassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
           echo "Response: $RESPONSE"
-          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "ChangeNotPermitted")'
+          # 6: ChangeNotPermitted
+          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == 6)'
 
       - name: LDAP Environment Failure Test
         run: |
@@ -136,8 +140,8 @@ jobs:
             -H "Content-Type: application/json" \
             -d '{"username": "testuser", "currentPassword": "testpassword", "newPassword": "NewPassword123!", "newPasswordVerify": "NewPassword123!"}')
           echo "Response: $RESPONSE"
-          # Expect LdapProblem or InvalidCredentials (if it fails to connect during bind)
-          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == "InvalidCredentials" or .errorCode == "LdapProblem")'
+          # Expect LdapProblem (8) or InvalidCredentials (4)
+          echo $RESPONSE | jq -e '.errors[] | select(.errorCode == 4 or .errorCode == 8)'
 
           echo "Restarting MokAPI..."
           docker start mokapi

--- a/src/Unosquare.PassCore.Web/appsettings.LdapTest.json
+++ b/src/Unosquare.PassCore.Web/appsettings.LdapTest.json
@@ -8,7 +8,9 @@
     "LdapHostnames": [ "localhost" ],
     "LdapPort": 389,
     "LdapUsername": "cn=admin,ou=people,dc=example,dc=com",
-    "LdapPassword": "adminpassword"
+    "LdapPassword": "adminpassword",
+    "RestrictedADGroups": [ "RestrictedGroup" ],
+    "AllowedADGroups": [ "AllowedGroup" ]
   },
   "ClientSettings": {
     "PasswordProviderOptions": {

--- a/src/Zyborg.PassCore.PasswordProvider.LDAP/LdapPasswordChangeProvider.cs
+++ b/src/Zyborg.PassCore.PasswordProvider.LDAP/LdapPasswordChangeProvider.cs
@@ -74,7 +74,7 @@ public class LdapPasswordChangeProvider : PasswordChangeProviderBase, IGroupMemb
                 _options.LdapSearchBase,
                 LdapConnection.ScopeSub,
                 searchFilter,
-                new[] { "memberOf" },
+                null,
                 false,
                 _searchConstraints);
 
@@ -82,7 +82,8 @@ public class LdapPasswordChangeProvider : PasswordChangeProviderBase, IGroupMemb
                 return Task.FromResult(false);
 
             var entry = search.Next();
-            var memberOf = entry.GetAttribute("memberOf");
+            var attrSet = entry.GetAttributeSet();
+            var memberOf = attrSet.ContainsKey("memberOf") ? attrSet["memberOf"] : null;
 
             if (memberOf == null)
                 return Task.FromResult(false);

--- a/src/Zyborg.PassCore.PasswordProvider.LDAP/LdapPasswordChangeProvider.cs
+++ b/src/Zyborg.PassCore.PasswordProvider.LDAP/LdapPasswordChangeProvider.cs
@@ -37,6 +37,33 @@ public class LdapPasswordChangeProvider : PasswordChangeProviderBase, IGroupMemb
         null,
         10);
 
+    private static readonly string[] RequiredAttributes =
+    {
+        "distinguishedName",
+        "sAMAccountName",
+        "memberOf",
+        "userPassword",
+        "unicodePwd"
+    };
+
+    private static readonly Action<ILogger, string, Exception?> LogLdapSearchStart =
+        LoggerMessage.Define<string>(
+            LogLevel.Debug,
+            new EventId(101, nameof(LogLdapSearchStart)),
+            "Starting LDAP search with filter: {Filter}");
+
+    private static readonly Action<ILogger, string, string, Exception?> LogLdapUserFound =
+        LoggerMessage.Define<string, string>(
+            LogLevel.Debug,
+            new EventId(102, nameof(LogLdapUserFound)),
+            "LDAP user found: {Username} (DN: {UserDN})");
+
+    private static readonly Action<ILogger, string, string, Exception?> LogLdapAttributeFound =
+        LoggerMessage.Define<string, string>(
+            LogLevel.Debug,
+            new EventId(103, nameof(LogLdapAttributeFound)),
+            "LDAP attribute found: {AttributeName} = {AttributeValue}");
+
     // TODO: is this related to https://github.com/dsbenghe/Novell.Directory.Ldap.NETStandard/issues/101 at all???
     // Had to mark this as nullable.
     private LdapRemoteCertificateValidationCallback? _ldapRemoteCertValidator;
@@ -69,12 +96,14 @@ public class LdapPasswordChangeProvider : PasswordChangeProviderBase, IGroupMemb
             var cleanUsername = CleaningUsername(username);
             var searchFilter = _options.LdapSearchFilter.Replace("{Username}", cleanUsername, StringComparison.Ordinal);
 
+            LogLdapSearchStart(Logger, searchFilter, null);
+
             using var ldap = BindToLdap();
             var search = ldap.Search(
                 _options.LdapSearchBase,
                 LdapConnection.ScopeSub,
                 searchFilter,
-                null,
+                RequiredAttributes,
                 false,
                 _searchConstraints);
 
@@ -82,6 +111,8 @@ public class LdapPasswordChangeProvider : PasswordChangeProviderBase, IGroupMemb
                 return Task.FromResult(false);
 
             var entry = search.Next();
+            LogLdapUserFound(Logger, username, entry.Dn, null);
+
             var attrSet = entry.GetAttributeSet();
 
             // Try to find memberOf attribute in a case-insensitive way
@@ -91,13 +122,22 @@ public class LdapPasswordChangeProvider : PasswordChangeProviderBase, IGroupMemb
             var memberOf = memberOfKey != null ? attrSet[memberOfKey] : null;
 
             if (memberOf == null)
+            {
+                Logger.LogDebug("Attribute 'memberOf' not found for user: {Username}", username);
                 return Task.FromResult(false);
+            }
+
+            foreach (var val in memberOf.StringValueArray)
+            {
+                LogLdapAttributeFound(Logger, "memberOf", val, null);
+            }
 
             return Task.FromResult(memberOf.StringValueArray.Any(g =>
                 g.Contains(groupName, StringComparison.OrdinalIgnoreCase)));
         }
-        catch (LdapException)
+        catch (LdapException ex)
         {
+            Logger.LogWarning(ex, "LDAP error during group membership check for user: {Username}", username);
             return Task.FromResult(false);
         }
     }
@@ -118,7 +158,7 @@ public class LdapPasswordChangeProvider : PasswordChangeProviderBase, IGroupMemb
                 _options.LdapSearchBase,
                 LdapConnection.ScopeSub,
                 searchFilter,
-                new[] { "distinguishedName" },
+                RequiredAttributes,
                 false,
                 _searchConstraints);
 

--- a/src/Zyborg.PassCore.PasswordProvider.LDAP/LdapPasswordChangeProvider.cs
+++ b/src/Zyborg.PassCore.PasswordProvider.LDAP/LdapPasswordChangeProvider.cs
@@ -83,7 +83,12 @@ public class LdapPasswordChangeProvider : PasswordChangeProviderBase, IGroupMemb
 
             var entry = search.Next();
             var attrSet = entry.GetAttributeSet();
-            var memberOf = attrSet.ContainsKey("memberOf") ? attrSet["memberOf"] : null;
+
+            // Try to find memberOf attribute in a case-insensitive way
+            var memberOfKey = attrSet.Keys.Cast<string>()
+                .FirstOrDefault(k => k.Equals("memberOf", StringComparison.OrdinalIgnoreCase));
+
+            var memberOf = memberOfKey != null ? attrSet[memberOfKey] : null;
 
             if (memberOf == null)
                 return Task.FromResult(false);

--- a/tests/mokapi/users.ldif
+++ b/tests/mokapi/users.ldif
@@ -26,7 +26,7 @@ cn: testuser
 sn: testuser
 userPassword: testpassword
 sAMAccountName: testuser
-memberOf: cn=AllowedGroup,ou=people,dc=example,dc=com
+memberOf: cn=AllowedGroup,ou=groups,dc=example,dc=com
 
 dn: cn=alloweduser,ou=people,dc=example,dc=com
 objectClass: top
@@ -37,7 +37,7 @@ cn: alloweduser
 sn: alloweduser
 userPassword: testpassword
 sAMAccountName: alloweduser
-memberOf: cn=AllowedGroup,ou=people,dc=example,dc=com
+memberOf: cn=AllowedGroup,ou=groups,dc=example,dc=com
 
 dn: cn=restricteduser,ou=people,dc=example,dc=com
 objectClass: top
@@ -48,7 +48,7 @@ cn: restricteduser
 sn: restricteduser
 userPassword: testpassword
 sAMAccountName: restricteduser
-memberOf: cn=RestrictedGroup,ou=people,dc=example,dc=com
+memberOf: cn=RestrictedGroup,ou=groups,dc=example,dc=com
 
 dn: cn=unlisteduser,ou=people,dc=example,dc=com
 objectClass: top
@@ -59,3 +59,21 @@ cn: unlisteduser
 sn: unlisteduser
 userPassword: testpassword
 sAMAccountName: unlisteduser
+
+dn: ou=groups,dc=example,dc=com
+objectClass: top
+objectClass: organizationalUnit
+ou: groups
+
+dn: cn=AllowedGroup,ou=groups,dc=example,dc=com
+objectClass: top
+objectClass: groupOfNames
+cn: AllowedGroup
+member: cn=testuser,ou=people,dc=example,dc=com
+member: cn=alloweduser,ou=people,dc=example,dc=com
+
+dn: cn=RestrictedGroup,ou=groups,dc=example,dc=com
+objectClass: top
+objectClass: groupOfNames
+cn: RestrictedGroup
+member: cn=restricteduser,ou=people,dc=example,dc=com

--- a/tests/mokapi/users.ldif
+++ b/tests/mokapi/users.ldif
@@ -26,6 +26,7 @@ cn: testuser
 sn: testuser
 userPassword: testpassword
 sAMAccountName: testuser
+memberOf: cn=AllowedGroup,ou=people,dc=example,dc=com
 
 dn: cn=alloweduser,ou=people,dc=example,dc=com
 objectClass: top

